feat(apps/whale-api): add updatemasternode indexer

imkven opened the pull request:
#### What this PR does / why we need it:

Add `UpdateMasternodeIndexer extends DfTxIndexer<UpdateMasternode>` to update the masternode index when the DfTx UpdaterMasternode transaction is created.

#### Which issue(s) does this PR fixes?:

Fixes part of #1842

marktanrj committed:
Add updateMasternode rpc (WIP)

marktanrj committed:
Add test cases

marktanrj committed:
Undo update config that was accidentally committed

marktanrj committed:
Undo update config that was accidentally committed

marktanrj committed:
Update testcases

marktanrj committed:
Merge branch 'main' into marktanrj/updatemasternode-rpc-typing

marktanrj committed:
Update image version

marktanrj committed:
Update image version

marktanrj committed:
Update image version

marktanrj committed:
add dftx (WIP)

imkven committed:
Add mapping

imkven committed:
WIP

marktanrj committed:
Resolve conflict

marktanrj committed:
Update image version

marktanrj committed:
Remove redundant code

marktanrj committed:
Add transaction builder tests (WIP)

marktanrj committed:
Fix format

shohamc1 committed:
feat(v3): bump to grandcentral and use `master-1d169a751` (#1841)

Bump testcontainers to `grandcentral` with `master-1d169a75`. Fixes all broken tests and known regression from this change.

Co-authored-by: Dilshan Madushanka <helloscoopz@gmail.com>

marktanrj committed:
Fix format bug

marktanrj committed:
Add tests for updating operator and reward address

shohamc1 committed:
chore(jellyfish-api-core): Change assert order in `getMiningInfo.test.ts` to match other tests (#1865)

<!--  Thanks for sending a pull request! -->

#### What this PR does / why we need it:
Uses to the assertion order used by all other tests. Uses a regex
pattern to match instead of `toContain`.

#### Additional comments?:

As requested in
https://github.com/JellyfishSDK/jellyfish/pull/1841#discussion_r1028115217

marktanrj committed:
Add tests

marktanrj committed:
Update image version

marktanrj committed:
Update tests

marktanrj committed:
Update test and dftx for removing reward address

marktanrj committed:
Minor edit

marktanrj committed:
Add tests for transaction builder

marktanrj committed:
Fix mistakes in comments

shohamc1 committed:
feat(jellyfish-api-core): defid v3 poolpair/poolswap internal changes (#1854)

<!--  Thanks for sending a pull request! -->

#### What this PR does / why we need it:

Add tests for new fixes on defid v3.

marktanrj committed:
Fix buggy code for UpdateMasternodeData

marktanrj committed:
Minor changes

marktanrj committed:
Add jellyfish transaction tests

shohamc1 committed:
fix(jellyfish-api-core): update testPoolSwap arguments (#1856)

<!--  Thanks for sending a pull request! -->

#### What this PR does / why we need it:

Exposes the path and verbose arguments on `testpoolswap`. Removes the
defunct UTXO argument.

marktanrj committed:
Undo rebase mess up

marktanrj committed:
Merge main

marktanrj committed:
Merge main

marktanrj committed:
Add more tests

imkven committed:
Merge remote-tracking branch 'origin/marktanrj/updatemasternode-rpc-typing' into kvenho/updatemasternode-whale-indexer

imkven committed:
Add updatemasternode indexer

imkven committed:
Add invalidate for update masternode

infinia-yzl commented on apps/whale-api/src/module.indexer/model/dftx/update.masternode.spec.ts:
```diff
@@ -0,0 +1,184 @@
+    const gotMasternodeForReward = await client.masternode.getMasternode(masternodeId)
+    // expect(gotMasternodeForReward[masternodeId]?.ownerAuthAddress).toStrictEqual(addressDest.utf8String) // @TODO(kvenho) Blockchain is not fix yet
+    expect(gotMasternodeForReward[masternodeId]?.operatorAuthAddress).toStrictEqual(addressDest.utf8String)
+    expect(gotMasternodeForReward[masternodeId]?.rewardAddress).toStrictEqual(addressDest.utf8String)
```
```suggestion
    const mnResult = await client.masternode.getMasternode(masternodeId)
    expect(mnResult[masternodeId]).toStrictEqual({
        ownerAuthAddress: addressDest.utf8String,
        operatorAuthAddress: addressDest.utf8String,
        ...
    })
```

imkven replied:
Resolved in this commit: https://github.com/JellyfishSDK/jellyfish/pull/1869/commits/e342f76485d7015b03804ea3b2033339fd8aeca7

imkven replied:
Just realise. I think testing value one by one is correct way to do, because if we write the test with ONE `toStrictEqual`, we need to include the WHOLE object. E.g.

```
expect(mnResult[masternodeId]).toStrictEqual({
      ownerAuthAddress: addressDest.utf8String,
      operatorAuthAddress: addressDest.utf8String,
      rewardAddress: addressDest.utf8String,
      creationHeight: 103,
      resignHeight: -1,
      resignTx: '0000000000000000000000000000000000000000000000000000000000000000',
      collateralTx: 'd1b9bfbf5a7a1beab499be8bb05654f61398e305a942ac6a1ff0362d8ed31b34', <-- And, this is different in every test
      state: 'ENABLED',
      mintedBlocks: 0,
      ownerIsMine: true,
      operatorIsMine: true,
      localMasternode: false,
      targetMultipliers: [1, 1]
    })
```

infinia-yzl replied:
If the collateral txid cannot be obtained deterministically, then just assert the type.
```js
expect(mnResult[masternodeId]).toStrictEqual({
      ownerAuthAddress: addressDest.utf8String,
      operatorAuthAddress: addressDest.utf8String,
      rewardAddress: addressDest.utf8String,
      creationHeight: 103,
      resignHeight: -1,
      resignTx: '0000000000000000000000000000000000000000000000000000000000000000',
      collateralTx: expect.stringMatching(/[0-f]{64}/),
      state: 'ENABLED',
      mintedBlocks: 0,
      ownerIsMine: true,
      operatorIsMine: true,
      localMasternode: false,
      targetMultipliers: [1, 1]
    })
```

infinia-yzl commented on apps/whale-api/src/module.indexer/model/dftx/update.masternode.spec.ts:
```diff
@@ -0,0 +1,184 @@
+    expect(gotMasternodeForReward[masternodeId]?.rewardAddress).toStrictEqual(addressDest.utf8String)
+
+    const updatedMasternode = await masternodeMapper.get(masternodeId)
+    expect(updatedMasternode?.operatorAddress).toStrictEqual(addressDest.utf8String)
```
Similar to comment above, assert the entire result.

imkven replied:
Resolved in this commit: https://github.com/JellyfishSDK/jellyfish/pull/1869/commits/18699e91abd69b99ab9942a77d28039ac4e8450c

imkven replied:
Reverted.

infinia-yzl commented on apps/whale-api/src/module.indexer/model/dftx/update.masternode.spec.ts:
```diff
@@ -0,0 +1,184 @@
+    const updatedMasternode = await masternodeMapper.get(masternodeId)
+    expect(updatedMasternode?.operatorAddress).toStrictEqual(addressDest.utf8String)
+    // expect(updatedMasternode?.ownerAddress).toStrictEqual(addressDest.utf8String) // @TODO(kvenho) Blockchain is not fix yet
+
```
Add more statements to generate more blocks to change the `status` and test against the different statuses.

imkven replied:
Resolved: https://github.com/JellyfishSDK/jellyfish/pull/1869/commits/b6faef6f75403329e98643857d24a13a32fcc3c8

infinia-yzl reviewed (changes requested):
Couple of comments.

imkven committed:
Update tests

imkven committed:
Update update.masternode.spec.ts

Resolve: https://github.com/JellyfishSDK/jellyfish/pull/1869#discussion_r1030597842

imkven committed:
Update update.masternode.spec.ts

Resolve: https://github.com/JellyfishSDK/jellyfish/pull/1869#discussion_r1030601781

marktanrj committed:
Update tests

imkven committed:
Update update.masternode.spec.ts

imkven committed:
Merge branch 'marktanrj/updatemasternode-dftx' into kvenho/updatemasternode-whale-indexer

imkven committed:
Update update.masternode.spec.ts

shohamc1 commented on packages/jellyfish-api-core/src/category/masternode.ts:
```diff
@@ -144,6 +144,24 @@ export class Masternode {
+   * @param {UTXO[]} [utxos = []] Array of specified utxos to spend.
+   * @param {string} [utxos.txid] The transaction id.
+   * @param {number} [utxos.vout] The output number.
+   * @return {Promise<string>} Resignation Transaction.
```
```suggestion
   * @return {Promise<string>} Transaction hash
```

shohamc1 commented on docs/node/CATEGORIES/11-masternode.md:
```diff
@@ -172,6 +172,23 @@ interface UTXO {
 
+## updatemasternode
+
+Creates (and submits to local node and network) a masternode update transaction which update the masternode operator addresses, spending the given inputs..
```
```suggestion
Creates a transaction to update masternode operator address, owner address, and/or reward address.
```

infinia-yzl committed:
Update packages/jellyfish-api-core/src/category/masternode.ts

Co-authored-by: Shoham Chakraborty <shhmchk@gmail.com>
Signed-off-by: Isaac Yong <dev@infinia.space>

infinia-yzl committed:
Update docs/node/CATEGORIES/11-masternode.md

Co-authored-by: Shoham Chakraborty <shhmchk@gmail.com>
Signed-off-by: Isaac Yong <dev@infinia.space>

marktanrj committed:
Add updatemasternode rpc

marktanrj committed:
Attempt to fix update owner address in masternode test (WIP)

marktanrj committed:
Update

marktanrj committed:
Fix update owner address test

imkven committed:
Merge branch 'marktanrj/updatemasternode-dftx' into kvenho/updatemasternode-whale-indexer

marktanrj committed:
Update tests

marktanrj committed:
Merge main

marktanrj committed:
Update tests to use object assertion

marktanrj committed:
Minor refactor and fix for test

marktanrj committed:
Add tests

marktanrj committed:
Add test for createDeFiTx

imkven committed:
Merge branch 'marktanrj/updatemasternode-dftx' into kvenho/updatemasternode-whale-indexer

marktanrj committed:
Update vout utxo amounts to be more deterministic

marktanrj committed:
Resolve PR comments

marktanrj committed:
Remove transaction builder code and shift into separate branch

marktanrj committed:
Merge branch 'main' into marktanrj/updatemasternode-dftx

marktanrj committed:
Merge branch 'marktanrj/updatemasternode-dftx' into kvenho/updatemasternode-whale-indexer

fuxingloh committed:
Merge branch 'main' into kvenho/updatemasternode-whale-indexer

fuxingloh commented on apps/whale-api/src/module.indexer/model/dftx/update.masternode.spec.ts:
```diff
@@ -0,0 +1,240 @@
+      rewardAddress: addressDest.utf8String,
+      creationHeight: 103,
+      resignHeight: -1,
+      resignTx: expect.stringMatching(/[0-f]{64}/),
```
Can we fix these alerts?

shohamc1 committed:
Fix regex warnings

fuxingloh commented on apps/whale-api/src/module.indexer/model/dftx/update.masternode.spec.ts:
```diff
@@ -85,8 +85,8 @@ describe('Update masternode', () => {
       resignHeight: -1,
-      resignTx: expect.stringMatching(/[0-f]{64}/),
-      collateralTx: expect.stringMatching(/[0-f]{64}/),
+      resignTx: expect.stringMatching(/[0-9A-Fa-f]{64}/),
```
```suggestion
      resignTx: expect.stringMatching(/[0-9a-f]{64}/),
```

I'm assuming it's lowercaps only?

shohamc1 replied:
Yup that's correct. Fixed in d5c83a5f

shohamc1 committed:
Only check for lowercase hex characters

fuxingloh commented on apps/whale-api/src/module.indexer/model/dftx/update.masternode.ts:
```diff
@@ -0,0 +1,157 @@
+    }
+  }
+
+  async indexStats (block: RawBlock, data: Masternode): Promise<void> {
```
Hmm, it won't make sense to `indexStats` for an update? Since statistically there isn't any changes for updates?

- count, no changes, updating doesn't change the amount of masternode?
- tvl, what's locked is still locked

imkven replied:
Make sense. Thanks.

imkven replied:
Have removed. https://github.com/JellyfishSDK/jellyfish/pull/1869/commits/3538564ab474abf0101096754bf329e8b839c195

fuxingloh commented on apps/whale-api/src/module.indexer/model/dftx/update.masternode.ts:
```diff
@@ -0,0 +1,157 @@
+        block: { hash: block.hash, height: block.height, medianTime: block.mediantime, time: block.time },
+        collateral: txn.vout[1].value.toFixed(8),
+        updateRecords
+      })
```
We shouldn't be replacing all the entries since it's an update, not a new creation.

fuxingloh replied:
Updating an existing record would look something like this.

https://github.com/JellyfishSDK/jellyfish/blob/bee44d910bf817e3c41d0117b0cca60b1d86dd55/apps/whale-api/src/module.indexer/model/dftx/resign.masternode.ts#L26-L30

imkven replied:
Updated. https://github.com/JellyfishSDK/jellyfish/pull/1869/commits/0b133048664f76c6894d7d983b6640366a132aca

fuxingloh commented on apps/whale-api/src/module.indexer/model/dftx/update.masternode.ts:
```diff
@@ -0,0 +1,157 @@
+
+  async invalidateBlockStart (block: RawBlock): Promise<void> {
+    await this.masternodeStatsMapper.delete(block.height)
+  }
```
```suggestion
```

imkven replied:
Have removed. https://github.com/JellyfishSDK/jellyfish/pull/1869/commits/c54546315e79414e0cfa3e911a20eb9df0e1d6ad

fuxingloh commented on apps/whale-api/src/module.indexer/model/dftx/update.masternode.ts:
```diff
@@ -0,0 +1,157 @@
+      count: x.weeks === lockStats.weeks ? (x.count - lockStats.count) : x.count,
+      tvl: x.weeks === lockStats.weeks ? new BigNumber(x.tvl).minus(lockStats.tvl).toFixed(8) : x.tvl
+    }))
+  }
```
```suggestion
```

imkven replied:
Have removed. https://github.com/JellyfishSDK/jellyfish/pull/1869/commits/3538564ab474abf0101096754bf329e8b839c195

imkven committed:
Remove indexStats and mapTimelockStats

imkven committed:
Remove invalidateBlockStart

imkven committed:
Fix update too much

imkven committed:
Merge branch 'main' into kvenho/updatemasternode-whale-indexer

imkven committed:
Merge branch 'main' into kvenho/updatemasternode-whale-indexer

imkven committed:
Merge branch 'main' into kvenho/updatemasternode-whale-indexer

imkven committed:
Merge branch 'main' into kvenho/updatemasternode-whale-indexer

helloscoopa commented on apps/whale-api/src/module.indexer/model/dftx/update.masternode.ts:
```diff
@@ -0,0 +1,107 @@
+
+  constructor (
+    private readonly masternodeMapper: MasternodeMapper,
+    private readonly masternodeStatsMapper: MasternodeStatsMapper,
```
This can be removed as unused.
```suggestion
```

Commit:
Update apps/whale-api/src/module.indexer/model/dftx/update.masternode.ts

Co-authored-by: Dilshan Madushanka <helloscoopz@gmail.com>
Signed-off-by: Kven Ho <kven@cakedefi.com>

helloscoopa commented on apps/whale-api/src/module.indexer/model/dftx/update.masternode.ts:
```diff
@@ -0,0 +1,106 @@
+import { RawBlock } from '../_abstract'
+import { Inject, Injectable } from '@nestjs/common'
+import { MasternodeMapper } from '../../../module.model/masternode'
+import { MasternodeStatsMapper } from '../../../module.model/masternode.stats'
```
Sorry didn't mention in the suggestion.

```suggestion
```

Commit:
Update apps/whale-api/src/module.indexer/model/dftx/update.masternode.ts

Co-authored-by: Dilshan Madushanka <helloscoopz@gmail.com>
Signed-off-by: Kven Ho <kven@cakedefi.com>

helloscoopa committed:
Merge branch 'main' into kvenho/updatemasternode-whale-indexer

marktanrj commented on apps/whale-api/src/module.indexer/model/dftx/update.masternode.ts:
```diff
@@ -0,0 +1,105 @@
+        ownerAddress: (ownerAddress !== null) ? ownerAddress : mn.ownerAddress,
+        operatorAddress: (operatorAddress !== null) ? operatorAddress : mn.operatorAddress,
+        updateRecords
+      })
```
thought: Not sure if we need to index reward address as well, since it is not in the `apps/whale-api/src/module.model/masternode.ts` model interface

fuxingloh replied:
Let's not then

marktanrj reviewed (dismissed):
LGTM

fuxingloh commented:
`docs/node/CATEGORIES/11-masternode.md` applied in https://github.com/JellyfishSDK/jellyfish/pull/1921

fuxingloh committed:
Merge branch 'main' into kvenho/updatemasternode-whale-indexer

Signed-off-by: Fuxing Loh <4266087+fuxingloh@users.noreply.github.com>

fuxingloh committed:
chore(whale-api): optimize UpdateMasternode to run without modifying CreateMasternode (#1920)

#### What this PR does / why we need it:

This utilizes the "forward indexing" design; it doesn't require
`whale-api` to be reindexed. It will automatically create a history
entry to `MasternodeMapper`. You can update your `whale-api` node
without this change.
